src/test/java/org/folio/rest/api/ItemEffectiveLocationTest.java

```diff
@@ -0,0 +1,468 @@
+    return itemsClient.create(JsonObject.mapFrom(item));
+  }
+
+  // is actually used by @Parameters(method = "parameters")
+  @SuppressWarnings("unused")
+  private List<String[]> parameters() {
+    List<String[]> holdingLocations = Arrays.asList(
+      new String[]{null, null},
+      new String[]{null, MAIN_LIBRARY_LOCATION},
+      new String[]{SECOND_FLOOR_LOCATION, null},
+      new String[]{SECOND_FLOOR_LOCATION, MAIN_LIBRARY_LOCATION}
+    );
+
+    List<String[]> itemLocations = Arrays.asList(
+      new String[]{null, null},
+      new String[]{null, THIRD_FLOOR_LOCATION},
+      new String[]{FOURTH_FLOOR_LOCATION, null},
+      new String[]{FOURTH_FLOOR_LOCATION, THIRD_FLOOR_LOCATION}
+    );
+
+    List<String[]> newItemLocations = Arrays.asList(
+      new String[]{null, null},
+      new String[]{null, THIRD_FLOOR_LOCATION},
+      new String[]{THIRD_FLOOR_LOCATION, null},
+      new String[]{THIRD_FLOOR_LOCATION, FOURTH_FLOOR_LOCATION},
+      new String[]{null, ANNEX_LIBRARY_LOCATION},
+      new String[]{ANNEX_LIBRARY_LOCATION, null},
+      new String[]{ANNEX_LIBRARY_LOCATION, ONLINE_LOCATION}
+    );
+
+    // Combine the locations
+    List<String[]> parameters = new ArrayList<>();
+    for (String[] holdingsLocation : holdingLocations) {
+      for (String[] itemLocation : itemLocations) {
+        for (String[] newItemLocation : newItemLocations) {
+
+          parameters.add(new String[]{
+            holdingsLocation[0], holdingsLocation[1],
+            itemLocation[0], itemLocation[1],
+            newItemLocation[1], newItemLocation[1]
+          });
+        }
+      }
+    }
+
+    return parameters;
+  }
+
+  private UUID lookupLocationByName(String name)
+    throws InterruptedException, ExecutionException, TimeoutException, MalformedURLException {
+    if (name == null) {
+      return null;
+    }
+
+    return locationsClient.getAll().stream()
+      .filter(json -> json.getString("name").equals(name))
+      .map(json -> UUID.fromString(json.getString("id")))
+      .findAny()
+      .orElse(null);
+  }
+
+  private void setPermanentTemporaryLocation(
+    JsonObject json, UUID permanentLocationId, UUID tempLocationId) {
+    String permanentLocationIdFieldName = "permanentLocationId";
+    String temporaryLocationIdFieldName = "temporaryLocationId";
+
+    json.remove(permanentLocationIdFieldName);
+    json.remove(temporaryLocationIdFieldName);
+
+    if (permanentLocationId != null) {
+      json.put(permanentLocationIdFieldName, permanentLocationId.toString());
+    }
+
+    if (tempLocationId != null) {
+      json.put(temporaryLocationIdFieldName, tempLocationId.toString());
+    }
+  }
+
+  private void assertEffectiveLocation(JsonObject item, UUID holdingPerm, UUID holdingTemp, UUID itemPerm, UUID itemTemp) {
+    UUID expectedEffectiveLocation = ObjectUtils
+      .firstNonNull(itemTemp, itemPerm, holdingTemp, holdingPerm);
+
+    if (expectedEffectiveLocation == null) {
```

Comment:
Under what circumstances might this happen? Could it be better represented by explicitly expecting this during those tests?

Reply:
@marcjohnson-kint, I'm not sure if this is a real case. UI says that `permanetLocationId` for a holding record is a required field, however API does not.

Reply:
If all four locations (permanent and temporary of item and holding) are not defined then there is no effective location. This may happen because they are optional. An item that is ordered but not received (and therefore cannot been loaned) does not need any location.